[RDY] Add nvim_set_keymap method to remote API

Yilin-Yang opened the pull request:
Meant to address #9136.

This would be my first contribution, so I'm still working to find my feet. I think that I've figured out the core of what I need to do (accept mappings from the user, parse them, and then pass them to `int do_map()`), but I'm unsure of a lot of the particulars.

Right now, I'm trying to figure out:

- What should the function signature be? (Should the user just pass in strings containing the "components" of the mapping? Pros: more flexible, extensible, and "natural", Cons: more complicated string parsing + related weird corner cases)
- How much input validation should we perform on the arguments the user provides? (We could infer that `:nnozfdxfdwse` is an attempt to `:nnoremap`, but should we still reject that as an invalid command?)
- How verbosely/granularly should we report errors to the user? (Okay to just return -1 no matter the error, or is it better to return, e.g. -2 for malformed arguments, etc.)

I'm also wondering if there are existing helper functions (beyond `get_map_mode`) on which I can offload much of the string parsing, since I'm not too satisfied with what I have right now. I'd appreciate any input or suggestions!

Yilin-Yang committed:
Write function stub for nvim_set_keymap

Yilin-Yang committed:
Add runscript for keymap test

Yilin-Yang committed:
Fix FUNC_API_SINCE, add test assertion

Fix so that we pass that one functional test about metadata. Add a false
assertion so that we can tell that the test is running.

Yilin-Yang committed:
Invoke nvim_set_keymap from the tests

Yilin-Yang committed:
Add/update test scripts

Yilin-Yang committed:
Write "draft" implementation of nvim_set_keymap

Yilin-Yang committed:
Start writing test cases for nvim_set_keymap

Yilin-Yang committed:
Start adding "true" functionality tests

Not yet passing.

Yilin-Yang committed:
Force to foreground when running in gdbserver

Using a tip taken from the neovim contributors wiki.

Yilin-Yang committed:
Fix the compilation error

Yilin-Yang committed:
Add space characters in concatenation

Yilin-Yang committed:
Add maptype arg to add_map function

Yilin-Yang committed:
Also return do_map's ret val in add_map

Yilin-Yang committed:
Pass basic nvim_set_keymap test

Yilin-Yang committed:
Fix style issues reported by clint.py

Yilin-Yang committed:
Revert "Also return do_map's ret val in add_map"

This reverts commit e400c70705484e6c733cb5c30aac2c834cb645be.

Yilin-Yang committed:
Revert "Add maptype arg to add_map function"

This reverts commit 4b9d86f5494a629e7f0c2312568fed5ed91d1167.

These changes turned out to be unnecessary; using do_map "directly" is
viable.

Yilin-Yang committed:
Start writing additional tests, not passing some

Yilin-Yang committed:
Set GDB breakpoint in nvim_set_keymap

Since that's what I need to debug.

Yilin-Yang committed:
Remove superfluous goto; add TODO to tests

- goto statement didn't do anything except strip (potentially useful)
error values from the call to `do_map()`.
- Getting |:unmap| to work is next on the list.

mhinz commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+  int is_noremap = STRNCMP(p + 1, "noremap", 7) == 0;
+
+  // "unnoremap"/etc. isn't a real command
+  if (is_noremap && is_unmap) { goto RETURN_FAILURE; }
```
No one-liners, please. :)

Yilin-Yang replied:
That's fair, haha! No worries, I'll fix that before I RFC!

mhinz commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+  }
+  int map_excl = STRNCMP(p, "map!", 4) == 0;
+  int mode_val = get_map_mode(&p, map_excl);
+  int is_noremap = STRNCMP(p + 1, "noremap", 7) == 0;
```
```suggestion
  bool is_noremap = STRNCMP(p + 1, "noremap", 7) == 0;
```

mhinz commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+
+  // convert mode shortname into struct mapblock's integer representation
+  char_u *p = (char_u *)map_cmd.data;
+  int is_unmap = map_cmd.size && p[0] == 'u';
```
```suggestion
  bool is_unmap = map_cmd.size && p[0] == 'u';
```

mhinz commented on src/nvim/tui/tui.c:
```diff
@@ -302,6 +302,12 @@ static void terminfo_start(UI *ui)
+
+  // force TUI to foreground when running in gdbserver
+  signal (SIGTTOU, SIG_IGN);
+  if (!tcsetpgrp(data->input.in_fd, getpid())) {
```
`tcsetpgrp()` is not available on Windows.

mhinz commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+///
+/// @returns Zero on success, nonzero on failure.
+Integer nvim_set_keymap(String map_cmd, String map_args,
+                        String lhs, String rhs)
```
You could add `Error *err` to the list of arguments. Then use `api_set_error()` later.

bfredl commented:
Very nice! We probably want an `options` dictionary for less fundamental args like `silent` or `nowait`. A map of string keys is easier to deal with than string concatenation and parsing. Likewise we would want to split up `map_cmd` into its components, `mode` and separate `nore` flag. Unmap should likely be a separate API function. Possibly, `iabbr` and `cabbr` could be regarded as `mode` values.

mhinz committed:
Apply suggestions from code review

Co-Authored-By: Yilin-Yang <yiliny@umich.edu>

Yilin-Yang committed:
Replace one-line if statements

Yilin-Yang committed:
Revert "Force to foreground when running in gdbserver"

This reverts commit 88f02b47742e61c7aabc8658a1b998eab121b4a7 and
53c5977941507988b67c8a05f1fed715ef9b8b66.

Remove gdbserver-specific code (wasn't useful for debugging, and so that
I won't forget to do this later)

Yilin-Yang committed:
Rewrite nvim_set_keymap to use new signature

- Pass options in a dictionary, to simplify string parsing.
- Use a statically allocated buffer for arguments, to avoid potential
memory leaks.
- Use api_set_error for reporting errors back to the user, as suggested
by mhinz.

TODO: testing

Yilin-Yang committed:
Concatenate lhs, rhs onto the arg string

- We need a dynamically allocated buffer for this, so do that.
- Add some error handling for when do_map fails

Yilin-Yang committed:
Attempt to gettext() for set_keymap err messages

Yilin-Yang committed:
Goto failure on do_map fail, return 0/-1

To report success to the user.

Yilin-Yang committed:
Revert "Attempt to gettext() for set_keymap err messages"

This reverts commit 42849f9eb7fe197d64de069fb7c898bb1ef88857.

Other api_set_error calls don't use gettext; does api_set_error handle
those by itself?

Yilin-Yang committed:
Remove gettext calls in do_map error handling

Yilin-Yang committed:
Start rewriting tests to use new signature

Yilin-Yang committed:
Fix error handling tests, some other test syntax

Yilin-Yang committed:
Pass unmap tests

Yilin-Yang committed:
Add todo, related to string validation

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,179 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
 
+/// Sets a |mapping| for the given mode.
+///
+/// Note that isn't possible to specify a target buffer when opts contains the
```
We would like a separate `nvim_buf_set_keymap` for this. Compare how `nvim_get_keymap` and `nvim_buf_get_keymap` is implemented. Most of the docs can stay at `nvim_set_keymap`, and the `nvim_buf` variant can just reference it.

Yilin-Yang replied:
While I would love to implement this, I don't think that I have the time to do so, at least for now. Would it be possible to split this off into a different issue/PR?

bfredl replied:
New API functions should ideally not hardcode `curbuf` usage. I don't think this should be so difficult to achieve. Make `do_map` (or the second half, if the split suggestion is followed) take `buf_T *buf` arg and use it everywhere instead of `curbuf` inside of the function.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,179 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+///               |:map!|). Unrecognized or empty values are treated like "m".
+/// @param  maptype   Whether to |:map|, |:unmap|, or |:noremap|, represented by
+///                   an empty string, "u", and "n", respectively.
+/// @param  opts  Optional parameters. Includes all |:map-arguments| as keys.
```
typically `opts` is the last arg (but before `Error`). Urrecognized keys should be an error.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,179 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+/// @returns  Zero on success, nonzero on failure.
+Integer nvim_set_keymap(String mode, String maptype, Dictionary opts,
+                        String lhs, String rhs, Error *err)
+  FUNC_API_SINCE(6)  // TODO(Yilin-Yang): make sure this is correct
```
CI will give an error if this is incorrect, don't worry :)

Yilin-Yang committed:
Fix style issues reported by clint.py

Yilin-Yang committed:
Use stricter input validation; move opts to end

Response to code review comments on GitHub.
- Move opts to the end of the arguments list, but before *err (better
consistency with other API functions)
- Throw errors when given unrecognized arguments (don't fail silently;
announce errors so that they can be fixed by developers)
- Silence compiler warnings from using variables as a format string

Yilin-Yang committed:
Tweak valid mode shortnames

For better consistency with maparg(). Use '!' to represent ':map!', and
allow the use of the empty string, ' ', and 'm' for ':map'.

Yilin-Yang committed:
Add additional tests

For edge cases, error checking, etc.

Having issues getting expect_err to work as expected.

Yilin-Yang committed:
Properly check for "<script>" in nvim_set_keymap

Yilin-Yang committed:
DRY the tests; print verbose testnames

Mimicking the tests for nvim_get_keymap, place the it() call inside the
for loop rather than the other way around, so we can see exactly which
inputs are failing, rather than knowing (vaguely) that at least one is
failing.

Reduce code duplication across the keymap tests.

Yilin-Yang committed:
Check for nonempty rhs in unmap commands

Also fix the looped unmap tests.

Yilin-Yang committed:
Test for empty RHS checking with pure whitespace

Yilin-Yang committed:
Add ascii_islinebreak function, strip_whitespace

We want to reject mappings that contain literal linebreak characters,
and which contain spaces "in the middle."

strip_whitespace strips leading and trailing whitespace, and also
validates the input string (intended to be LHS, RHS.)

Yilin-Yang committed:
Fix sign error in strip_whitespace

Yilin-Yang committed:
Fix edge case handling in strip_whitespace

Errors would occur if both the first and last nonwhitespace character
were "on the same side" of the midpoint, since the
traversal-from-both-ends wouldn't cross over that midpoint.

Yilin-Yang committed:
Validate LHS/RHS for presence of whitespace

Since interrupting whitespace may break the call to do_map() (since it
might parse lhs as containing an lhs AND an rhs), and since literal
linebreak characters in map commands aren't allowed.

Yilin-Yang committed:
Add more through testing of special characters

Yilin-Yang committed:
Fix style errors reported by clint.py

Yilin-Yang committed:
Add additional error handling tests for whitespace

Yilin-Yang committed:
Modify String to_strip through a pointer

The old implementation/usage of strip_whitespace would copy-construct
the given String, twice. Because of this, the `size` member was not
being updated correctly.

Also, correctly handle the case where the given string is nothing but
whitespace.

Yilin-Yang committed:
Update set_keymap docs for use with gen_vimdocs.py

Yilin-Yang committed:
Run gen_vimdocs.py to pick up nvim_set_keymap

Yilin-Yang committed:
Fix clint.py style errors in helpers.c

Yilin-Yang commented:
I think that I'm close to done; aside from needing to update `doc/api.txt` (and rebasing + force-pushing, etc.), I feel satisfied with what we have right now. Are there any other potential changes that stand out?

Also, thanks so much for your help, @mhinz and @bfredl! Receiving so much useful feedback so quickly came as a very pleasant surprise, and it made working on this PR much less difficult than it might have been otherwise.

Yilin-Yang committed:
Minor syntax fix

Yilin-Yang committed:
Add more thorough tests of map-arguments

bfredl commented on gdb_start.sh:
```diff
@@ -0,0 +1,2 @@
+target remote localhost:6666
```
don't include this file in the PR. (can use `git rm --cached` to remove it from the branch while keeping the physical file).

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,225 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+  char *err_arg = NULL;  // argument for the error message format string
+  ErrorType err_type = kErrorTypeNone;
+
+  // make sure that lhs and rhs aren't purely whitespace, and don't contain
```
`rhs` and `lhs` should ideally allow whitespace, this is really just a parsing issue of `do_map`.  A strategy could be to split `do_map` after the parsing, probaly after the `if (maptype == 1 !haskey) {` condition. Then this API function could skip the whitespace parsing entirely, and send in lhs/rhs with whitespace in it (which then means the same as `<space>`) . Though maybe `\n` should still be forbidden, it litteraly is the same as `<c-j>` but many users would probably expect it to be mean `<cr>`

Yilin-Yang committed:
Remove test scripts from git tracking

Yilin-Yang committed:
getchar.c: Write doxygen for int do_map

In preparation for refactoring (splitting do_map into multiple funtions,
adding an overload that takes a buffer parameter).

Also change `abbrev` -> `is_abbrev`, and make it a bool, to better
reflect its usage. (Existing references to do_map only set `abbrev` to 0
or 1.)

Yilin-Yang committed:
getchar.c: Add buf_do_map function

Exactly the same as the old do_map, but replace the hardcoded references
to the global `buf_T* curbuf` with a function parameter so that we can
invoke it from an nvim_buf_set_keymap function.

Yilin-Yang committed:
getchar: Create new map_arguments struct

So that a string of map arguments can be parsed into a more useful, more
portable data structure.

Yilin-Yang committed:
Parse out lhs and rhs in str_to_mapargs

Yilin-Yang committed:
Update docs for map_arguments

Yilin-Yang committed:
getchar: Start splitting buf_do_map

- Start refactoring buf_do_map to use MapArguments, rather than parsing
everything out of a single string.
- Not yet working.

Yilin-Yang committed:
Correctly set noremap val when setting script map

Yilin-Yang committed:
keymap_spec.lua: Test handling of weird characters

Yilin-Yang committed:
vim.c: Update lhs, rhs len after replace_termcodes

Fixes a bug in which replace_termcodes changes the length of lhs or rhs,
but the later search through the mappings/abbreviations hashtables
still uses the old length value. This would cause the search to fail
erroneously and throw 'E31: No such mapping' errors or 'E24: No such
abbreviation' errors.

Yilin-Yang committed:
keymap_spec.lua: Remove "reject whitespace" tests

Being able to include literal whitespace in the LHS and RHS is desired
behavior.

Yilin-Yang committed:
getchar.c: replace_termcodes in str_to_mapargs

It makes sense to do this; str_to_mapargs is, intuitively, supposed to
take a "raw" command string and parse it into a totally "do_map-ready"
struct.

Yilin-Yang committed:
keymap_spec.lua: Use literal whitespace in tests

Since mappings can now contain literal space characters, newlines, etc.
update test cases to expect the appropriate values for LHS, RHS, etc.

Yilin-Yang committed:
api/vim.c: Bypass do_map call in nvim_set_keymap

So that, e.g. literal whitespace characters can be passed through
"as-is."

Yilin-Yang committed:
helpers.c: Parse set_keymap opts in a helper func

Yilin-Yang committed:
keymap_spec: Test error-checking <buffer> option

Yilin-Yang committed:
api/vim.c: Parse keymap opts with helper

Also, error-check to make sure that the user did not set <buffer> to
true.

Yilin-Yang committed:
api/buffer.c: Write untested nvim_buf_set_keymap

Reimplement nvim_set_keymap as a wrapped call to nvim_buf_set_keymap.

Yilin-Yang committed:
api/vim.c: Remove superfluous char buffer allocs

Yilin-Yang committed:
api/buffer.c: Fix clint.py style errors

bfredl commented on src/nvim/api/buffer.c:
```diff
@@ -575,6 +576,172 @@ ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
 
+/// Like |nvim_set_keymap|, but for a specific buffer.
+///
+/// If not provided in {opts}, |<buffer>| is assumed to be true. If provided
```
We don't need the 'buffer' key at all. if `nvim_buf_set_keymap` is used, the mapping is buffer-local, and with `nvim_set_keymap` the mapping is global.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1245,6 +1245,62 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
+    goto FAIL_WITH_MESSAGE;
+  }
+
+  return nvim_buf_set_keymap(0, mode, maptype, lhs, rhs, opts, err);
```
Pass -1 as first arg to signal global behavior, just like `nvim_get_commands` (User code can't pass in -1).

Yilin-Yang committed:
getchar.c: Fix clint.py errors in buf_do_map_expl

Yilin-Yang committed:
getchar.c: Fix clint.py errs in str_to_mapargs

Yilin-Yang committed:
api/helpers.c: Check union type before read

Avoid undefined behavior when the given value is not actually a bool.

Yilin-Yang committed:
api: Remove 'buffer' as an option for set_keymap

Since it's now made redundant by the presence of nvim_buf_set_keymap.

Yilin-Yang committed:
buffer_spec.lua: Start testing buf_set_keymap

Yilin-Yang committed:
api/buffer.c: find_buffer_by_handle in set_keymap

Yilin-Yang committed:
keymap_spec.lua: Test expr, noremap, nowait

Add functional tests for <expr>, noremap, and <nowait> to make sure that
these arguments are handled properly.

Yilin-Yang committed:
keymap_spec.lua: Test rhs, lhs of MAXMAPLEN

Yilin-Yang committed:
getchar.c: Use allocated buffer for rhs in MapArgs

Since the MAXMAPLEN limit does not apply to the RHS of a mapping (or
else an RHS that calls a really long autoload function from a plugin
would be incorrectly rejected as being too long), use an allocated
buffer for RHS rather than a static buffer of length MAXMAPLEN + 1.

Yilin-Yang committed:
keymap_spec.lua: Add tests for RHS = <nop>

Yilin-Yang committed:
getchar.c: Pull "repl. MapArg termcodes" into func

The "preprocessing code" surrounding the replace_termcodes calls needs
to invoke replace_termcodes, and also check if RHS is equal to "<Nop>".
To reduce code duplication, factor this out into a helper function.

Also add an rhs_is_noop flag to MapArguments; buf_do_map_explicit
expects an empty {rhs} string for "<Nop>", but also needs to distinguish
that from something like ":map lhs<cr>" where no {rhs} was provided.

Yilin-Yang committed:
api: Regen API docs, make various style fixes

Also remove redundant option-parsing code from nvim_set_keymap (that's
now handled by nvim_buf_set_keymap.)

Yilin-Yang committed:
test: Fix testlint errors

bfredl commented on src/nvim/api/buffer.c:
```diff
@@ -575,6 +576,150 @@ ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
+    }  // switch
+  }
+  assert(!(is_unmap && is_noremap));
+  if (!is_unmap && (parsed_args.rhs_len == 0 && !parsed_args.rhs_is_noop)) {
```
This was a restriction of `:map` ex command parsing, where empty argument is impossible to type. For API function there is no reason to disallow empty rhs.

Yilin-Yang committed:
api: Treat empty {rhs} like <Nop>, add more tests

Yilin-Yang committed:
api/buffer.c: Throw err_msg if assert doesn't trip

For when nvim is compiled with -DNDEBUG.

justinmk commented on src/nvim/ascii.h:
```diff
@@ -168,4 +172,13 @@ static inline bool ascii_isspace(int c)
+/// one of \n or \r.
+///
+/// @see {ascii_isdigit}
+static inline bool ascii_islinebreak(int c)
```
does not seem to be used

justinmk commented:
@Yilin-Yang This is really nicely done, thank you!

> * How much input validation should we perform on the arguments the user provides? (We could infer that `:nnozfdxfdwse` is an attempt to `:nnoremap`, but should we still reject that as an invalid command?)

The API typically is "rigid", avoid "smart" behavior. "Smart" (unpredictable) behavior is already covered by VimL, and we don't need more of that in the API.

> * How verbosely/granularly should we report errors to the user? (Okay to just return -1 no matter the error, or is it better to return, e.g. -2 for malformed arguments, etc.)

The convention for API functions is to call `api_set_error` on the `Error` parameter, this  manifests as an exception in all clients. I see that `nvim_set_keymap` returns an error code, but I would suggest `void` return type (if there is nothing else meaningful to return). Rather the `Error` message text should reflect granular error case. In #6150 we decided that API error messages will not be translated. (We may also introduce new error "types" in the future, but the error messages will be reliable in any case.)

justinmk commented on src/nvim/getchar.h:
```diff
@@ -23,6 +24,35 @@ typedef enum {
+///
+/// The <special> argument has no effect on mappings and is excluded from this
+/// struct declaration.
+struct map_arguments {
```
/// @see mapblock_T

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -575,6 +576,157 @@ ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
+  }
+
+  if (mode.size > 1) {
+    err_msg = "Given shortname is too long: %s";
```
"shortname is too long: %s"

("given" is implied)

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -575,6 +576,157 @@ ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
+      // get_map_mode will treat "unrecognized" mode shortnames like "map"
+      // if it does, and the given shortname wasn't "m" or " ", then error
+      if (STRNCMP(p, "m", 2) && STRNCMP(p, " ", 2)) {
+        err_msg = "Unrecognized mode shortname: %s";
```
"Invalid mode shortname: %s";

We try to use the same language everywhere. Instead of "unrecognized", we usually say "invalid".

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -575,6 +576,157 @@ ArrayOf(Dictionary) nvim_buf_get_keymap(Buffer buffer, String mode, Error *err)
+  }
+
+  if (parsed_args.lhs_len == 0) {
+    err_msg = "Must give nonempty LHS!";
```
"Invalid (empty) LHS"

we don't use "!" in messages :)

justinmk commented on test/functional/api/buffer_spec.lua:
```diff
@@ -435,6 +437,96 @@ describe('api/buf', function()
     end)
   end)
 
+  describe('nvim_buf_set_keymap', function()
```
if it's not too much trouble could you move this to `api/keymap_spec.lua`? That's where the tests for  `nvim_[buf_]_get_keymap` live.

justinmk commented on test/functional/api/keymap_spec.lua:
```diff
@@ -308,3 +311,394 @@ describe('nvim_get_keymap', function()
+    local keys = 'i123456'
+    for c in string.gmatch(keys, '.') do
+      feed(c)
+      sleep(5)
```
`helpers.wait()` is a more deterministic and less-expensive way to do this.

justinmk commented on src/nvim/api/private/helpers.c:
```diff
@@ -744,6 +744,92 @@ String ga_take_string(garray_T *ga)
+        break;
+    }  // switch
+    if (!was_valid_opt) {
+      err_msg = "Unrecognized option in nvim_set_keymap: %s";
```
```suggestion
      err_msg = "Invalid key: %s";
```

Error messages do not restate the method name.

bfredl commented:
I think API here is starting to get really good, but I have an issue about the naming. `nvim_get_keymap` returns _all_ mappings in a mode, so it seems "a keymap" refers to a set of mappings (`help mbyte-keymap` also uses the word in the same way). Should the function be called `nvim_add_mapping` and `nvim_del_mapping` ? 

At least bundling `unmap` and `noremap` in the same arg as is done now seems a bit weird. `nore` is essentially a modifier (like `<expr>`)  that became part of the command name because reasons, while `unmap` is like a separate command (it doesn't take the same args and options etc).

justinmk commented:
> Should the function be called `nvim_add_mapping` and `nvim_del_mapping` ?

I think `keymap` is fine. It was chosen for `nvim_set_keymap` to avoid confusion with a dictionary "map". I would rather have get/set "duals" than  `get_keymap` and `set_mapping`, it is less discoverable.

bfredl commented:
Ok, but I would stil opt for `nvim_set_keymap` and `nvim_del_keymap` then, it is essentially analogous to `set_var` and `del_var`.

Yilin-Yang committed:
Update with suggestions from code review

- Remove unused `ascii_islinebreak` function (`strip_whitespace`
function in which it was used no longer exists)
- Reword error messages for better consistency with other API functions
- Update tests to reflect reworded error messages
- Move `nvim_buf_set_keymap` tests to `keymap_spec.lua`
(keep everything grouped in one file, like with `nvim_[buf_]get_keymap`)
- Minor documentation tweaks.

Yilin-Yang committed:
Change set_keymap return type to void

justinmk commented:
Will wait for `nvim_del_keymap` and then this is ready to merge.

Yilin-Yang committed:
Add nvim_[buf_]del_keymap functions

Yilin-Yang committed:
Remove maptype argument from set_keymap signature

Was a holdover from `int do_map`'s original implementation, where
noremap/unmap/neither was signified with an int.

unmap now has its own separate function, so noremap can go in the
options dict. This lets us remove the maptype argument entirely, which
also makes testing easier, since we don't need to test unmapping with
both del_keymap AND set_keymap with maptype = 'u'.

Yilin-Yang commented:
I went ahead and added `nvim_[buf_]del_keymap` functions. Since that obviates the need for `set_keymap` to accept `maptype = "u"`, `maptype` would then only be useful for setting `noremap`, so I went ahead with @bredl's suggestion to make `noremap` an opt and then got rid of `maptype` entirely. That was a good suggestion; originally I was just following the signature of `do_map`, but breaking from that does make for a cleaner interface.

> regarding the new MapArgument type: did you consider adding fields to mapblock_T instead?

Also a good suggestion that I didn't think of while coding 😅. I do think it's better to have a separate type, though, for clarity's sake: since `buf_do_map_explicit` (and the original `do_map`) sometimes allocates/inserts a new `mapblock_T` into the global hashtables and sometimes doesn't (if it can reuse an existing one), it might be unclear who's left with "ownership" of the struct if it were passed into `buf_do_map_explicit` as a parameter.


Thanks for your guys' help! (and also don't be afraid to ping me if there are issues in the code i just pushed!)

justinmk commented:
Merged. Thank you @Yilin-Yang , this is really appreciated. 

I squashed into 1 commit because most of the intermediate commits were not meaningful. For big/complex PRs, it's good to do it as you did: small commits without force-push. Then before merge it might make sense to rebase into a few meaningful commits--but for this PR I didn't think that would be worth your time. I preserved  commit messages where relevant.

I will make a followup PR with a few minor nitpicks.
